flink-runtime/src/main/java/org/apache/flink/runtime/clusterframework/BootstrapTools.java

```diff
@@ -621,23 +621,38 @@ public Config getAkkaConfig() {
 	 *
 	 * @param baseConfig The base configuration.
 	 * @param targetConfig The target configuration.
-	 * @return Dynamic properties as string, separated by whitespace.
+	 * @return Dynamic properties as string array.
 	 */
-	public static String getDynamicProperties(Configuration baseConfig, Configuration targetConfig) {
+	public static String[] getDynamicProperties(Configuration baseConfig, Configuration targetConfig) {
 
-		String[] newAddedConfigs = targetConfig.keySet().stream().flatMap(
+		return targetConfig.keySet().stream().flatMap(
 			(String key) -> {
 				final String baseValue = baseConfig.getString(ConfigOptions.key(key).stringType().noDefaultValue());
 				final String targetValue = targetConfig.getString(ConfigOptions.key(key).stringType().noDefaultValue());
 
 				if (!baseConfig.keySet().contains(key) || !baseValue.equals(targetValue)) {
 					return Stream.of("-" + CommandLineOptions.DYNAMIC_PROPERTY_OPTION.getOpt() + key +
-						CommandLineOptions.DYNAMIC_PROPERTY_OPTION.getValueSeparator() + targetValue);
+						CommandLineOptions.DYNAMIC_PROPERTY_OPTION.getValueSeparator() + escapeDynamicPropertyValue(targetValue));
 				} else {
 					return Stream.empty();
 				}
 			})
 			.toArray(String[]::new);
-		return String.join(" ", newAddedConfigs);
+	}
+
+	/**
+	 * Escape all the dynamic property values. Each value will be surrounded with single quotes. This works for all chars
+	 * except single quote itself. To escape the single quote, close the quoting before it, insert the escaped single
+	 * quote, and then re-open the quoting. For example, the value is my'value and the escaped value is 'my'\''value'.
+	 *
+	 * @param value value to be escaped
+	 * @return escaped value
+	 */
+	private static String escapeDynamicPropertyValue(String value) {
```

Comment:
I think this method could be tested explicitly. We should also make sure that we add corner cases (`'foobar`, `foo''bar`, etc.)

Reply:
I will add a separate test for encoding and try to cover the corner cases.